File: Java/Problem3.java
package Java;

import java.lang.Math;

public class Problem3 {

    public String shortestPalindrome(String s) {
        if (s.isEmpty()) {
            return s;
        }

        String sMirror = new StringBuffer(s).reverse().toString();

        int step = 0;
        while (true) {
            String test = s;
            String temp = new StringBuffer(test).substring(0, step).toString();

            test = temp + test;

            boolean isOdd = test.length() % 2 == 1;
            int pivot = isOdd ? Math.ceilDiv(test.length(), 2) : test.length() / 2;
            pivot -= 1;

            String firstPart = new StringBuffer(test).substring(0, pivot).toString();
            String firstReversed = new StringBuffer(firstPart).reverse().toString();
            String secondPart = new StringBuffer(test).substring(pivot).toString();

            System.out.println(firstReversed);
            System.out.println(secondPart);

            if (firstPart.equalsIgnoreCase(secondPart)) {
                return test;
            }

            step++;

        }

    }

}

File: Python/Problem3.py
class Problem3:
    def shortestPalindrome(self, s: str) -> str:
        if s.isspace() or s == "":
            return s
        s_mirror: str = s[::-1]

        step: int = 0

        while True:
            test: str = s
            temp = s_mirror[:step]
            test = temp + test

            is_odd: bool = len(test) % 2 == 1
            plus: int = 1 if is_odd else 0

            if test[: len(test) // 2 + plus][::-1] == test[len(test) // 2 :]:
                return test

            step += 1


test: str = "abbacd"
expected: str = "dcabbacd"

p3: Problem3 = Problem3()


print(p3.shortestPalindrome(test))
